src/Ampersand/Input/PreProcessor.hs

```diff
@@ -0,0 +1,110 @@
+module Ampersand.Input.PreProcessor (
+      preProcess
+    , PreProcDefine
+) where
+
+import Data.List
+import Data.String
+import Data.Char
+import Data.Either
+import Data.Maybe
+import Prelude
+
+type PreProcDefine = String
+
+preProcess :: [PreProcDefine] -> String -> String
+preProcess defs = block2file defs True . file2block
+ 
+-- Run the full parser
+file2block :: String -> Block
```

Comment:
I guess your code would become more maintainable if you would use another type here.  The intent is to take the contents of a file and turn it into a Block. If I understand what you envision, I think you see each line as a token, and then do the parsing of these tokens to a Block. 
so why not create a parser for a Block:

```
pBlock :: AmpParser Block
pBlock = many pBlockElem

pBlockElem :: AmpParser BlockElem
pBlockElem = pCommonLine
         <|> pIfBlock

pIfBlock :: AmpParser BlockElem
 etc. 
```

For this to work, you have to add some extra Lexemes (see LexerToken.hs, line 25 and further)